_posts/2019-07-31-neuron2.md

```diff
@@ -0,0 +1,276 @@
+compensate. The only adjustments this neuron can make are to its weights, and we want it to be making adjustments based
+on its error. To calculate this error, we subtract the predicted value from the expected value, which gives us
+these possible error values:
+<table class="ui celled table">
+    <thead>
+    <tr>
+        <th>Desired</th>
+        <th>Predicted</th>
+        <th>Error</th>
+    </tr>
+    </thead>
+    <tbody>
+    <tr>
+        <td>1</td>
+        <td>1</td>
+        <td>0</td>
+    </tr>
+    <tr>
+        <td>1</td>
+        <td>-1</td>
+        <td>2</td>
+    </tr>
+    <tr>
+        <td>-1</td>
+        <td>-1</td>
+        <td>0</td>
+    </tr>
+    <tr>
+        <td>-1</td>
+        <td>1</td>
+        <td>-2</td>
+    </tr>
+    </tbody>
+</table>
+
+We then take these error values, multiply them by our training rate (I'll come back to this), and then multiply that value
+by each input to get the values we need to steer each weight. I think this makes more sense to actually see in code, so
+here goes.
+
+#### Adjusting weights
+We'll start with writing a function that will handle adjusting the weights. 
+```python
+class Neuron:
+    # ...
+    def _adjust_weights(self, inputs, error):
+        for i in range(self.number_of_inputs):
+            self.weights[i] += inputs[i] * error * self.training_rate
+```
+So here, we take every input, multiply it by the overall error, then by the training rate, then we add that to that
+input's respective weight.
+
+#### Training Rate
+Basically, a training rate is how hard you steer the neuron. Very high training rates will have a tendency to oscillate
+back and forth over the optimal weight, each time overshooting it, and trying to correct back, and subsequently
+overshooting again. To solve this, we steer by only small amounts at a time. This does mean it can take longer to land
+on the optimal weight. So there is a balance to how hard we are going to steer our weights, and that value will probably
+differ from dataset to dataset. I have chosen, somewhat arbitrarily, a training rate of 0.1, set at the neuron's 
+initialization.
+```python
+class Neuron:
+    def __init__(self, number_of_inputs, weights=None, training_rate=0.1):
+        self.number_of_inputs = number_of_inputs
+        self.weights = weights if weights is not None else self._random_weights()
+        self.training_rate = training_rate
+```
+This makes it easy to play around with training rates without having to change my base neuron code each time.
+We have the ability to adjust our weights, now it's time to flex those neuron muscles... brains have muscles, right? 
+No... moving on...
+
+#### Put in reps
+We are going to make a function called, predictably, `train`. This function could work a couple of ways, it could take
+in a single set of inputs, the single expected output, make a prediction, and then adjust its weights, only doing this
```

Comment:
Could maybe use bullets here to split up this large paragraph?

--------

We are going to make a function called, predictably, train. This function could work a couple of ways:

 * it could take in a single set of inputs, the single expected output, make a prediction, and then adjust its weights, only doing this once per call
 * or, we could pass it our whole set of training data and let it train on all of it. 

For this neuron, I’m choosing to go with a step training process, feeding it the training data one piece at a time. This will make my implementation of my neuron slightly more cumbersome, since I’ll have to handle the data looping there, but it will allow me to do some interesting things when we get to testing.